File: apps/api/src/agentforge_api/models/agent.py
# apps/api/src/agentforge_api/models/agent.py

"""Agent domain models."""

from enum import Enum
from typing import Annotated, Any, NewType

from pydantic import BaseModel, Field


AgentId = NewType("AgentId", str)


class DataType(str, Enum):
    """Primitive types supported for agent inputs/outputs."""
    
    STRING = "string"
    NUMBER = "number"
    BOOLEAN = "boolean"
    OBJECT = "object"
    ARRAY = "array"


class PortSchema(BaseModel, frozen=True):
    """Schema for a single input or output port."""
    
    name: str
    type: DataType
    required: bool = True
    description: str = ""


class AgentInputSchema(BaseModel, frozen=True):
    """Definition of an agent's input requirements."""
    
    ports: list[PortSchema] = Field(default_factory=list)


class AgentOutputSchema(BaseModel, frozen=True):
    """Definition of an agent's output structure."""
    
    ports: list[PortSchema] = Field(default_factory=list)


class AgentCategory(str, Enum):
    """Agent category for organization and filtering."""
    
    LLM = "llm"              # Language model invocation
    RETRIEVAL = "retrieval"  # RAG, vector search
    TRANSFORM = "transform"  # Data transformation
    INTEGRATION = "integration"  # External API calls
    LOGIC = "logic"          # Branching, conditionals


class RetryPolicy(BaseModel, frozen=True):
    """Retry behavior for agent execution."""
    
    max_retries: Annotated[int, Field(ge=0)] = 3
    backoff_ms: Annotated[int, Field(ge=0)] = 1000
    backoff_multiplier: Annotated[float, Field(ge=1.0)] = 2.0


class AgentDefinition(BaseModel, frozen=True):
    """
    Complete definition of an agent available in the system.
    
    This is a template/blueprint, not an instance.
    Nodes reference agents by AgentId.
    """
    
    id: Annotated[str, Field(description="Unique agent identifier")]
    name: str
    description: str = ""
    category: AgentCategory
    
    input_schema: AgentInputSchema = Field(default_factory=AgentInputSchema)
    output_schema: AgentOutputSchema = Field(default_factory=AgentOutputSchema)
    
    default_config: dict[str, Any] = Field(default_factory=dict)
    cacheable: bool = True
    retry_policy: RetryPolicy = Field(default_factory=RetryPolicy)
    
    @property
    def agent_id(self) -> AgentId:
        """Return typed AgentId."""
        return AgentId(self.id)

File: packages/contracts/src/types/agent.ts
// src/types/agent.ts

/**
 * Unique identifier for an agent definition.
 */
export type AgentId = string & { readonly __brand: "AgentId" };

/**
 * Primitive types supported for agent inputs/outputs.
 */
export type DataType = "string" | "number" | "boolean" | "object" | "array";

/**
 * Schema for a single input or output port.
 */
export interface PortSchema {
  readonly name: string;
  readonly type: DataType;
  readonly required: boolean;
  readonly description: string;
}

/**
 * Definition of an agent's input requirements.
 */
export interface AgentInputSchema {
  readonly ports: readonly PortSchema[];
}

/**
 * Definition of an agent's output structure.
 */
export interface AgentOutputSchema {
  readonly ports: readonly PortSchema[];
}

/**
 * Agent category for organization and filtering.
 */
export type AgentCategory =
  | "llm" // Language model invocation
  | "retrieval" // RAG, vector search
  | "transform" // Data transformation
  | "integration" // External API calls
  | "logic"; // Branching, conditionals

/**
 * Complete definition of an agent available in the system.
 *
 * This is a template/blueprint, not an instance.
 * Nodes reference agents by AgentId.
 */
export interface AgentDefinition {
  readonly id: AgentId;
  readonly name: string;
  readonly description: string;
  readonly category: AgentCategory;

  /** Schema for inputs this agent accepts */
  readonly inputSchema: AgentInputSchema;

  /** Schema for outputs this agent produces */
  readonly outputSchema: AgentOutputSchema;

  /** Default configuration values */
  readonly defaultConfig: Record<string, unknown>;

  /** Whether this agent's output can be cached */
  readonly cacheable: boolean;

  /** Default retry policy */
  readonly retryPolicy: RetryPolicy;
}

/**
 * Retry behavior for agent execution.
 */
export interface RetryPolicy {
  readonly maxRetries: number;
  readonly backoffMs: number;
  readonly backoffMultiplier: number;
}
